xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(version)
+              .setNode(node)
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    /**
+     * Sends a DiscoveryRequest with the given information as an ACK. Updates the latest accepted
+     * version for the corresponding resource type.
+     */
+    private void sendAckRequest(String typeUrl, Collection<String> resourceNames,
+        String versionInfo, String nonce) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        ldsVersion = versionInfo;
+        ldsRespNonce = nonce;
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        rdsVersion = versionInfo;
+        rdsRespNonce = nonce;
+      }
+      // TODO(chengyuanzhang): cases for CDS/EDS.
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node)
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .build();
+      requestWriter.onNext(request);
+    }
+
+    /**
+     * Sends a DiscoveryRequest with the given information as an NACK. NACK takes the previous
+     * accepted version.
+     */
+    private void sendNackRequest(String typeUrl, Collection<String> resourceNames, String nonce,
+        String message) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      String versionInfo = "";
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        versionInfo = ldsVersion;
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        versionInfo = rdsVersion;
+      }
+      // TODO(chengyuanzhang): cases for CDS/EDS.
+      DiscoveryRequest request =
+          DiscoveryRequest
+              .newBuilder()
+              .setVersionInfo(versionInfo)
+              .setNode(node)
+              .addAllResourceNames(resourceNames)
+              .setTypeUrl(typeUrl)
+              .setResponseNonce(nonce)
+              .setErrorDetail(com.google.rpc.Status.newBuilder().setMessage(message))
+              .build();
+      requestWriter.onNext(request);
+    }
+  }
+
+  /**
+   * Returns {@code true} iff {@code hostName} matches the domain name {@code pattern} with
+   * case-insensitive.
+   *
+   * <p>Wildcard pattern rules:
+   * <ol>
+   * <li>A single asterisk (*) matches any domain.</li>
+   * <li>Asterisk (*) is only permitted in the left-most or the right-most part of the pattern,
+   *     but not both.</li>
+   * </ol>
+   */
+  @VisibleForTesting
+  static boolean matchHostName(String hostName, String pattern) {
+    // Basic sanity checks
+    if (hostName.length() == 0 || hostName.startsWith(".") || hostName.endsWith(".")) {
```

Comment:
`matchHostName("", "")` returning `false` is counter-intuitive, not documented, without a valid usecase. 

`matchHostName(".", ".")` returning `false` is also counter-intuitive, and not documented,

Reply:
From the perspective of a utility helper method, they are cases we want to exclude obviously. From this method's perspective, we couldn't say if `matchHostName("", "")` should return `true` or `false`. But I am more leaning to a specification implicitly indicates that it should return `false` for `hostName` and `pattern` being very wrong.

Anyway, I don't bother keeping deleting them. (IMO, this helper method looks more robust and more correct with these basic checks).

Reply:
Replaced with checks.